text/0000-enum-variant-types.md

```diff
@@ -0,0 +1,203 @@
+However, pattern-matching on the variant allows irrefutable matches on the particular variant. In
+effect, `Variant` is only relevant to type checking/inference and the matching logic.
+
+Constructors of variants, as well as pattern-matching on particular enum variants, are now
+inferred to have variant types, rather than enum types.
+
+```rust
+enum Sum {
+    A(u8),
+    B,
+    C,
+}
+
+let x = Sum::A(5); // x: Sum::A
+let Sum::A(y) = x; // ok, y = 5
+
+fn sum_match(s: Sum) {
+    match s {
+        a @ Sum::A(_) => {
+            let x = a; // ok, a: Sum::A
+        }
+        b @ Sum::B => {
+            // b: Sum::B
+        }
+        c @ Sum::C => {
+            // c: Sum::C
+        }
+    }
+}
+```
+
+In essence, a value of a variant is considered to be a value of the enclosing `enum` in every matter
+but pattern-matching.
+
+Explicitly casting to the `enum` type forgets the variant information.
+
+```rust
+let x: Sum = Sum::A(5); // x: Sum
+let Sum::A(y) = x; // error: refutable match
+```
+
+In all cases, the most specific type (i.e. the variant type if possible) is chosen by the type
+inference. However, this is entirely backwards-compatible, because `Variant` acts as `Adt` except in
+cases that were previously invalid (i.e. pattern-matching, where the extra typing information was
+previously unknown).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- The loose distinction between the `enum` type and its variant types could be confusing to those unfamiliar with variant types. Error messages might specifically mention a variant type, which could
+be at odds with expectations. However, since they generally behave identically, this should not
+prove to be a significant problem.
+- As variant types need to include generic parameter information that is not necessarily included in
+their definitions, it will be necessary to include explicit type annotations more often than is
+typical. Although this is unfortunate, it is necessary to preserve all the desirable behaviour of
+variant types described here: namely complete backwards-compatibility precise type inference
+(e.g. allowing `x` in `let x = Sum::A;` to have type `Sum::A` without explicit type annotations).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The advantages of this approach are:
+- It naturally allows variants to be treated as types, intuitively.
+- It doesn't require explicit type annotations to reap the benefits of variant types.
+- As variant types and enum types are represented identically, there are no coercion costs.
+- It doesn't require type fallback.
+- It doesn't require value-tracking or complex type system additions such as refinement types.
+
+One obvious alternative is to represent variant types differently to enum types and then coerce them
+when used as an enum. This could potentially reduce memory overhead for smaller variants
```

Comment:
That makes it worse than having dedicated `struct`s - is that something good? discuss!